File: src/bourse/step_sim/agents/random_agent.py
"""
Agent that places random uniformly sampled orders
"""
import typing

import numpy as np

from bourse import core
from bourse.step_sim.agents import BaseAgent, BaseNumpyAgent, InstructionArrays


class RandomAgent(BaseAgent):
    """
    Agent that place randomly sampled orders

    Agent that places orders on a random side with
    random volume and price within given ranges, where
    values are sampled from uniform distributions.

    Warnings
    --------
    The behaviour of this agent is not intended to
    represent any 'realistic' behaviour, so should
    really be used for testing or benchmarking.
    """

    def __init__(
        self,
        i: int,
        activity_rate: float,
        tick_range: typing.Tuple[int, int],
        vol_range: typing.Tuple[int, int],
        tick_size: int,
    ):
        """
        Initialise a RandomAgent

        Parameters
        ----------
        i: int
            Id of the agent
        activity_rate: float
            Activity rate of the agent(i.e. the
            probability the agent is active each step).
        tick_range: tuple[int, int]
            Tick range to sample from.
        vol_range: tuple[int, int]
            Volume range to sample from.
        tick_size: int
            Size of a market tick
        """
        self.i = i
        self.activity_rate = activity_rate
        self.tick_range = tick_range
        self.vol_range = vol_range
        self.tick_size = tick_size
        self.order_id = None

    def update(self, rng: np.random.Generator, env: core.StepEnv):
        """
        Cancel a live order or place a new random one

        If the agent is active this step it will:

        - If it has an active order on the market it
          will submit an instruction to cancel it.
        - If it has no active orders it will place a new
          order on a random side with random volume and price.

        Parameters
        ----------
        rng: numpy.random.Generator
            Numpy random generator.
        env: bourse.core.StepEnv
            Discrete event simulation environment.
        """
        p = rng.random()

        if p < self.activity_rate:
            # If an order is live then cancel it
            if self.order_id is not None and env.order_status(self.order_id) == 1:
                env.cancel_order(self.order_id)
                self.order_id = None
            # Otherwise place a new random order
            else:
                tick = rng.integers(*self.tick_range)
                vol = rng.integers(*self.vol_range)
                side = bool(rng.choice([True, False]))
                self.order_id = env.place_order(
                    side, vol, self.i, price=tick * self.tick_size
                )


class NumpyRandomAgents(BaseNumpyAgent):
    """
    Simple agent set that places random orders via Numpy arrays

    Agents that place random orders each step of the simulation, new
    orders are returned as a tuple of Numpy arrays. These orders
    can then be submitted to a discrete event environment using
    :py:meth:`bourse.core.StepEnv.submit_limit_order_array`.

    This agent type is designed to represent a group of agents all
    placing individual orders at each step (rather than a single agent).
    """

    def __init__(
        self,
        n_agents: int,
        tick_range: typing.Tuple[int, int],
        vol_range: typing.Tuple[int, int],
        tick_size: int,
    ):
        """
        Initialise NumpyRandomAgents

        Parameters
        ----------
        n_agents: int
            Number of agents in the set
        tick_range: tuple[int, int]
            Tick range to sample from.
        vol_range: tuple[int, int]
            Volume range to sample from.
        tick_size: int
            Size of a market tick
        """
        self.n_agents = n_agents
        self.tick_range = tick_range
        self.vol_range = vol_range
        self.tick_size = tick_size

    def update(
        self, rng: np.random.Generator, level_2_data: np.ndarray
    ) -> InstructionArrays:
        """
        Update the agents, sampling new orders to place

        Parameters
        ----------
        rng: numpy.random.Generator
            Numpy random generator.
        level_2_data: bourse.core.StepEnv
            Level-2 market data

        Returns
        -------
        tuple
            Tuple containing new order instructions
        """
        sides = rng.choice([True, False], size=self.n_agents).astype(bool)
        vols = rng.integers(*self.tick_range, size=self.n_agents, dtype=np.uint32)
        ids = np.arange(self.n_agents, dtype=np.uint32)
        prices = (
            rng.integers(*self.tick_range, size=self.n_agents, dtype=np.uint32)
            * self.tick_size
        )

        return (
            np.ones(self.n_agents, dtype=np.uint32),
            sides,
            vols,
            ids,
            prices,
            np.zeros(self.n_agents, dtype=np.uint64),
        )

File: crates/step_sim/src/agents/random_agent.rs
use super::Agent;
use crate::types::{OrderId, Price, Side, Status, TraderId, Vol};
use crate::Env;
use rand::seq::SliceRandom;
use rand::Rng;
use rand::RngCore;

/// Agents that place orders with uniformly sampled parameters
///
/// A set of agents that place orders on a random side with
/// random volume and price within a given range. Sides,
/// volumes and prices are uniformly sampled.
///
/// Each step each agent:
///
/// - Are randomly activated with a given activity rate
/// - If active and they have an active order on the
///   market they try to cancel that order
/// - If they don't have an active order then they
///   place a new order on a random side, with a
///   random volume and price
///
/// <div id="doc-warning-1" class="warning">
/// The behaviour of this agent is not based on anything
/// 'realistic' and should probably only be used for
/// testing and benchmarking.
/// </div>
///
/// # Examples
///
/// ```
/// use bourse_de::agents::{Agent, RandomAgents, Agents};
/// use bourse_de::{sim_runner, Env};
///
/// #[derive(Agents)]
/// struct SimAgents {
///     pub a: RandomAgents,
/// }
///
/// let mut env = Env::new(0, 1, 1_000_000, true);
///
/// let mut agents = SimAgents {
///     a: RandomAgents::new(10, (40, 60), (10, 20), 2, 0.8),
/// };
///
/// sim_runner(&mut env, &mut agents, 101, 10, false);
/// ```
pub struct RandomAgents {
    orders: Vec<Option<OrderId>>,
    tick_range: (Price, Price),
    vol_range: (Vol, Vol),
    tick_size: Price,
    activity_rate: f32,
}

impl RandomAgents {
    /// Initialise a set of random agents
    ///
    /// # Arguments
    ///
    /// - `n_agents` - Number of agents in the set
    /// - `tick_range` - Range of ticks to place orders over
    /// - `vol_range` - Order volume range to sample from
    /// - `tick_size` - Market tick size
    /// - `activity_rate` - Agent activity rate
    ///
    pub fn new(
        n_agents: usize,
        tick_range: (Price, Price),
        vol_range: (Vol, Vol),
        tick_size: Price,
        activity_rate: f32,
    ) -> Self {
        Self {
            orders: vec![None; n_agents],
            tick_range,
            vol_range,
            tick_size,
            activity_rate,
        }
    }
}

impl Agent for RandomAgents {
    fn update<R: RngCore>(&mut self, env: &mut Env, rng: &mut R) {
        let new_orders: Vec<Option<OrderId>> = self
            .orders
            .iter_mut()
            .enumerate()
            .map(|(n, i)| {
                let p = rng.gen::<f32>();

                match p < self.activity_rate {
                    true => {
                        if (i.is_some()) && (env.order_status(i.unwrap()) == Status::Active) {
                            env.cancel_order(i.unwrap());
                            None
                        } else {
                            let side = [Side::Ask, Side::Bid].choose(rng).unwrap();
                            let tick = rng.gen_range(self.tick_range.0..self.tick_range.1);
                            let vol = rng.gen_range(self.vol_range.0..self.vol_range.1);
                            Some(
                                env.place_order(
                                    *side,
                                    vol,
                                    TraderId::try_from(n).unwrap(),
                                    Some(tick * self.tick_size),
                                )
                                .unwrap(),
                            )
                        }
                    }
                    false => *i,
                }
            })
            .collect();

        self.orders = new_orders;
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use bourse_book::types::Event;
    use rand::SeedableRng;
    use rand_xoshiro::Xoroshiro128StarStar;

    #[test]
    fn test_activity_rate() {
        let mut env = Env::new(0, 1, 1000, true);
        let mut rng = Xoroshiro128StarStar::seed_from_u64(101);

        let mut agents = RandomAgents::new(2, (10, 20), (20, 30), 1, 0.0);

        agents.update(&mut env, &mut rng);
        assert!(env.get_transactions().len() == 0);

        agents.activity_rate = 1.0;
        agents.update(&mut env, &mut rng);
        assert!(env.get_transactions().len() == 2);
    }

    #[test]
    fn test_order_place_then_cancel() {
        let mut env = Env::new(0, 1, 1000, true);
        let mut rng = Xoroshiro128StarStar::seed_from_u64(101);

        let mut agents = RandomAgents::new(1, (10, 20), (20, 30), 1, 1.0);

        agents.update(&mut env, &mut rng);
        assert!(env.get_transactions().len() == 1);
        assert!(matches!(env.get_transactions()[0], Event::New { .. }));
        assert!(agents.orders == vec![Some(0)]);

        env.step(&mut rng);

        agents.update(&mut env, &mut rng);
        assert!(env.get_transactions().len() == 1);
        assert!(matches!(
            env.get_transactions()[0],
            Event::Cancellation { .. }
        ));

        env.step(&mut rng);

        agents.update(&mut env, &mut rng);
        assert!(env.get_transactions().len() == 1);
        assert!(matches!(env.get_transactions()[0], Event::New { .. }));
        assert!(agents.orders == vec![Some(1)]);
    }
}
